### PR TITLE
Send HostnameDiscoveryEvent from Agent

### LINKED ISSUE #2569: Update hostname on `Machine` when registration an `Agent`
## Describe the bug
The first time that the `IMachineRepository` is initiated it is upserting a machine with a hostname( just for Island). All the agents that get registered do not collect the hostname of the machine that they run. Which is why no other machine has a hostname.

## To Reproduce
1. Configure the Monkey to exploit any machine
2. Check to see that any exploited monkey will only display the first IP from their network interfaces

## Expected behavior
We should be collecting hostname from Machines when possible.

## Screenshots
image

## Machine version (please complete the following information):
 - OS: Windows or Linux

### PR DESCRIPTION
**Pull request by shreyamalviya**
# What does this PR do?

Fixes #2569 

## PR Checklist
* [x] Have you added an explanation of what your changes do and why you'd like to include them?
* [x] Is the TravisCI build passing?
* [ ] ~Was the CHANGELOG.md updated to reflect the changes?~
* [ ] ~Was the documentation framework updated to reflect the changes?~
* [x] Have you checked that you haven't introduced any duplicate code?

## Testing Checklist

* [ ] ~Added relevant unit tests?~
* [x] Have you successfully tested your changes locally? Elaborate:
    > Tested by running on GCP
* [x] If applicable, add screenshots or log transcripts of the feature working

image

### CODE REVIEW

File: monkey/infection_monkey/monkey.py [hunk 1]
```diff
@@ -215,6 +218,15 @@ def _discover_os(self):
         )
         self._agent_event_queue.publish(event)
 
+    def _discover_hostname(self):
+        event = HostnameDiscoveryEvent(
+            source=self._agent_id,
+            timestamp=time.time(),
+            tags={T1592_ATTACK_TECHNIQUE_TAG},
+            hostname=gethostname(),
```
Comment (ilija-lazoroski) [inline monkey/infection_monkey/monkey.py:226]: I think we should use `platform.node()`. It seems that is more efficient then `socket.gethostname`.

image

Comment (shreyamalviya) [reply monkey/infection_monkey/monkey.py:226]: It's in microseconds which means it won't be noticeable. I don't care about which one we use but another comment in [that thread](https://stackoverflow.com/a/4271873) suggests that `platform.node()` may be unreliable in some (rare) cases.
image

File: monkey/infection_monkey/monkey.py [hunk 2]
```diff
@@ -8,6 +8,7 @@
 from ipaddress import IPv4Interface
 from itertools import chain
 from pathlib import Path, WindowsPath
+from socket import gethostname
```
Comment (mssalvatore) [inline monkey/infection_monkey/monkey.py:11]: Let's wrap `gethostname()` in a new function: `utils.environment.get_hostname()`.

File: monkey/infection_monkey/monkey.py [hunk 3]
```diff
@@ -215,6 +218,15 @@ def _discover_os(self):
         )
         self._agent_event_queue.publish(event)
 
+    def _discover_hostname(self):
+        event = HostnameDiscoveryEvent(
+            source=self._agent_id,
+            timestamp=time.time(),
+            tags={T1592_ATTACK_TECHNIQUE_TAG},
```
Comment (mssalvatore) [inline monkey/infection_monkey/monkey.py:225]: I don't think [1592](https://attack.mitre.org/techniques/T1592/) is the correct attack technique (for OS discovery, either). It's for collecting information about remote systems, not local systems.

[T1082](https://attack.mitre.org/techniques/T1082/) is a better choice for both hostname and OS discovery.

**Review by mssalvatore — CHANGES_REQUESTED** [review]

File: monkey/infection_monkey/monkey.py (hunk 3 above)
Comment (VakarisZ) [reply monkey/infection_monkey/monkey.py]: [1592](https://attack.mitre.org/techniques/T1592/) is not part of PingScanEvent though 🤔

**Review by mssalvatore — APPROVED** [review]

**Pull request merged**